### PR TITLE
Emformer with conv module and scaling mechanism

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR implements Emformer augmented with convolution module. It is based on `pruned_transducer_stateless4`, which uses scaling mechanism and saves averaged model. 

Since this PR is to be merged into `master` branch, for the sake of simplicity, I directly copied files from my branch https://github.com/yaozengwei/icefall/tree/emformer_conv, which contains some changes in the `streaming` branch. Detailed commit history can be found on https://github.com/yaozengwei/icefall/commits/emformer_conv. 

The Emformer implementation is based on the torchaudio:
- emformer attention: https://github.com/pytorch/audio/blob/main/torchaudio/models/emformer.py 
- convolution module: https://github.com/pytorch/audio/blob/main/torchaudio/prototype/models/conv_emformer.py

Experimental results will be uploaded soon.
TODO:
- [x] Pack the convolution cache and attention states into a unified state
- [x] Test using torchscript to export model
- [x] Integrate streaming inference functions in the `streaming` branch

### CODE REVIEW

**Comment by yaozengwei** [comment]
When training on train-clean-100, and decoding with greedy search, I get the following result:

- epoch-30-avg-10, **9.09** on test-clean and **24.68** on test-other

The training command is:
```bash
./conv_emformer_transducer_stateless/train.py \                                                                                                                                               --world-size 2 \
  --num-epochs 30 \
  --start-epoch 1 \
  --exp-dir conv_emformer_transducer_stateless/exp-100
  --full-libri 0 \
  --max-duration 250 \
  --num-encoder-layers 12 \
  --chunk-length 32 \  
  --cnn-module-kernel 31 \
  --left-context-length 32 \
  --right-context-length 8 \
  --memory-size 32
```
The decoding command is:
```bash
./conv_emformer_transducer_stateless/decode.py \                                                                                                                                              
  --epoch 30 \
  --avg 10 \                                                                                                                                                                                
  --exp-dir conv_emformer_transducer_stateless/exp-100 \                                                                                                                                                                        
  --num-encoder-layers 12 \                                                                                                                                                                   
  --chunk-length 32 \                                                                                                                                                                         
  --cnn-module-kernel 31 \                                                                                                                                                                    
  --left-context-length 32 \                                                                                                                                                                  
  --right-context-length 8 \                                                                                                                                                                  
  --memory-size 32 \                                                                                                                                                                          
  --decoding-method greedy_search \                                                                                                                                                           
  --use-averaged-model True 
```
Note: The lengths used in `chunk-length`, `left-context-length`, and `right-context-length` are all before subsampling.

**Comment by yaozengwei** [comment]
When training on full librispeech, and decoding with greedy search, I get the following result:

- epoch-30-avg-10, **3.86** on test-clean and **9.8** on test-other

The training command is:
```bash
./conv_emformer_transducer_stateless/train.py \                                                                                                                                               
  --world-size 6 \
  --num-epochs 30 \
  --start-epoch 1 \
  --exp-dir conv_emformer_transducer_stateless/exp-full
  --full-libri 1 \
  --max-duration 300 \
  --num-encoder-layers 12 \
  --chunk-length 32 \  
  --cnn-module-kernel 31 \
  --left-context-length 32 \
  --right-context-length 8 \
  --memory-size 32
```
The decoding command is:
```bash
./conv_emformer_transducer_stateless/decode.py \                                                                                                                                              
  --epoch 30 \
  --avg 10 \                                                                                                                                                                                
  --exp-dir conv_emformer_transducer_stateless/exp-full \                                                                                                                                                                        
  --num-encoder-layers 12 \                                                                                                                                                                   
  --chunk-length 32 \                                                                                                                                                                         
  --cnn-module-kernel 31 \                                                                                                                                                                    
  --left-context-length 32 \                                                                                                                                                                  
  --right-context-length 8 \                                                                                                                                                                  
  --memory-size 32 \                                                                                                                                                                          
  --decoding-method greedy_search \                                                                                                                                                           
  --use-averaged-model True 
```
Note: The lengths used in `chunk-length`, `left-context-length`, and `right-context-length` are all before subsampling.

**Comment by danpovey** [comment]
@yaozengwei it's good to have that WER result, but do we have any comparisons for that?

**Comment by yaozengwei** [comment]
> > @yaozengwei it's good to have that WER result, but do we have any comparisons for that?
> 
> I had trained my implemented Emformer without conv module or scaling mechanism before, which has 18 layers, and I got the result:
> 
> * epoch-30-avg-10, **4.45** on test-clean and **11.44** on test-other
> 
> Fungjun had trained the Emformer using the torchaudio code (without conv module, https://github.com/pytorch/audio/blob/main/torchaudio/models/emformer.py), with ` --num-encoder-layers 18, --left-context-length 128, --segment-length 8, --right-context-length 4` (the length is also before subsampling), and got the result:
> 
> * epoch 29--avg 6 --decoding_method modified_beam_search, **4.58** on test-clean and **11.54** on test-other
>   See [Use Emformer as RNN-T encoder. #278](https://github.com/k2-fsa/icefall/pull/278).

In addition, I have trained the emfomer with convolution module, but without scaling mechanism, 
* epoch-30--avg-10 --decoding_method greedy_search, --use-averaged-model **False**,  **3.98** on test-clean and **10.34** on test-other

For comparison, the **current** moded with scaling mechanism, 
* epoch-30-avg-10 --decoding_method greedy_search, --use-averaged-model **False**, **4.0** on test-clean and **9.84** on test-other

**Comment by csukuangfj** [comment]
To get a WER below 3.0 for test-clean that is comparable to the one listed in the Emformer paper:
https://arxiv.org/pdf/2010.10759.pdf

Screen Shot 2022-05-31 at 17 47 08

We may need to use
- More layers (24)
- A larger right context size (32)

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/decode.py [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+    device = next(model.parameters()).device
+    feature = batch["inputs"]
+    assert feature.ndim == 3
+
+    feature = feature.to(device)
+    # at entry, feature is (N, T, C)
+
+    supervisions = batch["supervisions"]
+    feature_lens = supervisions["num_frames"].to(device)
+
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/decode.py:264]: Please add tail paddings here

https://github.com/k2-fsa/icefall/blob/1b909d917826112f3d53cfedaee16362acb7be91/egs/librispeech/ASR/pruned_stateless_emformer_rnnt2/decode.py#L268-L273

See #390 and #384

Comment (yaozengwei) [reply egs/librispeech/ASR/conv_emformer_transducer_stateless/decode.py:264]: OK

**Comment by yaozengwei** [comment]
In `conv_emformer_transducer_stateless/streaming_decode.py`, it supports streaming decoding for Emformer model in a chunk-by-chunk fashion. Main efforts including:
- mask out zero initial states, which is needed for batch streaming decoding;
- drop the first and last frame after convolution subsampling.
Now the streaming decoding results are almost same as in simulated streaming decoding:

In real streaming decoding (decoding methods, test-clean & test-other):
-  greedy_seach  **3.64 & 9.65** 
-  modified_beam_search  **3.54  & 9.46**
-  fast_beam_search **3.58 & 9.5**

In simulated streaming decoding (decoding methods, test-clean & test-other):
-  greedy_seach  **3.62 & 9.6** 
-  modified_beam_search  **3.55  & 9.39**
-  fast_beam_search **3.6 & 9.41**

**Comment by csukuangfj** [comment]
Please add the label `ready` when you think it's ready to merge.

**Comment by yaozengwei** [comment]
> In `conv_emformer_transducer_stateless/streaming_decode.py`, it supports streaming decoding for Emformer model in a chunk-by-chunk fashion. Main efforts including:
> 
> * mask out zero initial states, which is needed for batch streaming decoding;
> * drop the first and last frame after convolution subsampling.
> 
> Now the streaming decoding results are almost same as in simulated streaming decoding:
> 
> 1. In real streaming decoding (decoding methods, test-clean & test-other):
> 
> * greedy_seach  **3.64 & 9.65**
> * modified_beam_search  **3.54  & 9.46**
> * fast_beam_search **3.58 & 9.5**
> 
> 2. In simulated streaming decoding (decoding methods, test-clean & test-other):
> 
> * greedy_seach  **3.62 & 9.6**
> * modified_beam_search  **3.55  & 9.39**
> * fast_beam_search **3.6 & 9.41**

The streaming decoding commands are as follows:

- greedy search

```
./conv_emformer_transducer_stateless/streaming_decode.py \                                                                                                                                                      
      --epoch 30 \                                                                                                                                                                                              
      --avg 10 \                                                                                                                                                                                                  
      --exp-dir conv_emformer_transducer_stateless/exp-960_num_layer-12_chunk-32_conv-31_left-32_right-8_memory-32 \                                                                                                
      --num-encoder-layers 12 \                                                                                                                                                                                     
      --chunk-length 32 \                                                                                                                                                                                           
      --cnn-module-kernel 31 \                                                                                                                                                                                      
      --left-context-length 32 \                                                                                                                                                                                    
      --right-context-length 8 \                                                                                                                                                                                    
      --memory-size 32 \                                                                                                                                                                                            
      --decoding-method greedy_search \                                                                                                                                                                             
      --use-averaged-model True \                                                                                                                                                                                   
      --num-decode-streams 2000
```

- modified beam search
```
./conv_emformer_transducer_stateless/streaming_decode.py \                                                                                                                                                      
      --epoch 30 \                                                                                                                                                                                              
      --avg 10 \                                                                                                                                                                                                  
      --exp-dir conv_emformer_transducer_stateless/exp-960_num_layer-12_chunk-32_conv-31_left-32_right-8_memory-32 \                                                                                                
      --num-encoder-layers 12 \                                                                                                                                                                                     
      --chunk-length 32 \                                                                                                                                                                                           
      --cnn-module-kernel 31 \                                                                                                                                                                                      
      --left-context-length 32 \                                                                                                                                                                                    
      --right-context-length 8 \                                                                                                                                                                                    
      --memory-size 32 \                                                                                                                                                                                            
      --decoding-method modified_beam_search \                                                                                                                                                                      
      --use-averaged-model True \                                                                                                                                                                                   
      --beam-size 4 \                                                                                                                                                                                               
      --num-decode-streams 200
```

- fast beam search
```
./conv_emformer_transducer_stateless/streaming_decode.py \                                                                                                                                                      
      --epoch 30 \                                                                                                                                                                                              
      --avg 10 \                                                                                                                                                                                                  
      --exp-dir conv_emformer_transducer_stateless/exp-960_num_layer-12_chunk-32_conv-31_left-32_right-8_memory-32 \                                                                                                
      --num-encoder-layers 12 \                                                                                                                                                                                     
      --chunk-length 32 \                                                                                                                                                                                           
      --cnn-module-kernel 31 \                                                                                                                                                                                      
      --left-context-length 32 \                                                                                                                                                                                    
      --right-context-length 8 \                                                                                                                                                                                    
      --memory-size 32 \                                                                                                                                                                                            
      --decoding-method fast_beam_search \                                                                                                                                                                          
      --use-averaged-model True \                                                                                                                                                                                   
      --beam 4 \                                                                                                                                                                                                    
      --max-contexts 4 \                                                                                                                                                                                            
      --max-states 8 \                                                                                                                                                                                              
      --num-decode-streams 2000
```

**Comment by yaozengwei** [comment]
> Please add the label `ready` when you think it's ready to merge.

OK, I will do it after testing using torchscript.

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py [hunk 2]
```diff
@@ -0,0 +1,980 @@
...
+    add_model_arguments(parser)
+
+    return parser
+
+
+def greedy_search(
+    model: nn.Module,
+    encoder_out: torch.Tensor,
+    streams: List[Stream],
+) -> List[List[int]]:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py:248]:
```suggestion
) -> None:
```
This function returns nothing.

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py [hunk 3]
```diff
@@ -0,0 +1,980 @@
...
+
+def fast_beam_search_one_best(
+    model: nn.Module,
+    streams: List[Stream],
+    encoder_out: torch.Tensor,
+    processed_lens: torch.Tensor,
+    beam: float,
+    max_states: int,
+    max_contexts: int,
+) -> List[List[int]]:
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py:430]:
```suggestion
):
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py [hunk 4]
```diff
@@ -0,0 +1,980 @@
...
+        A tensor of shape (N,) containing the number of processed frames
+        in `encoder_out` before padding.
+      beam:
+        Beam value, similar to the beam used in Kaldi..
+      max_states:
+        Max states per stream per frame.
+      max_contexts:
+        Max contexts pre stream per frame.
+    Returns:
+      Return the decoded result.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py:453]:
```suggestion
      Return None.
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py [hunk 5]
```diff
@@ -0,0 +1,980 @@
...
+
+    params = get_params()
+    params.update(vars(args))
+
+    assert params.decoding_method in (
+        "greedy_search",
+        "fast_beam_search",
+        "modified_beam_search",
+    )
+    # Note: params.decoding_method is currently not used.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/streaming_decode.py:812]:
```suggestion
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
+import torch
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py:1]:
```suggestion
#!/usr/bin/env python3

import torch
```

And please also add copyright info.

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def test_state_stack_unstack():
+    from emformer import Emformer, stack_states, unstack_states
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py:5]: Please put it outside of the test function.

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    num_features = 80
+    chunk_length = 32
+    encoder_dim = 512
+    num_encoder_layers = 2
+    kernel_size = 31
+    left_context_length = 32
+    right_context_length = 8
+    memory_size = 32
+    batch_size = 2
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py:15]: Please also test `batch_size=1`.

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/train.py [hunk 9]
```diff
@@ -0,0 +1,1136 @@
...
+def compute_loss(
+    params: AttributeDict,
+    model: Union[nn.Module, DDP],
+    sp: spm.SentencePieceProcessor,
+    batch: dict,
+    is_training: bool,
+    warmup: float = 1.0,
+) -> Tuple[Tensor, MetricsTracker]:
+    """
+    Compute CTC loss given the model and its inputs.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/train.py:622]:
```suggestion
    Compute RNN-T loss given the model and its inputs.
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py [hunk 10]
```diff
@@ -0,0 +1,1886 @@
...
+from icefall.utils import make_pad_mask
+
+
+LOG_EPSILON = math.log(1e-10)
+
+
+def unstack_states(
+    states: Tuple[List[List[torch.Tensor]], List[torch.Tensor]]
+) -> List[Tuple[List[List[torch.Tensor]], List[torch.Tensor]]]:
+    # TODO: modify doc
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py:45]: What are you planning to do?

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py [hunk 11]
```diff
@@ -0,0 +1,1886 @@
...
+    def __init__(
+        self,
+        chunk_length: int,
+        right_context_length: int,
+        channels: int,
+        kernel_size: int,
+        bias: bool = True,
+    ) -> None:
+        """Construct an ConvolutionModule object."""
+        super(ConvolutionModule, self).__init__()
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py:165]:
```suggestion
        super().__init__()
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py [hunk 12]
```diff
@@ -0,0 +1,1886 @@
...
+        # kernerl_size should be a odd number for 'SAME' padding
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py:166]:
```suggestion
        # kernerl_size should be an odd number for 'SAME' padding
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py [hunk 13]
```diff
@@ -0,0 +1,1886 @@
...
+        assert (kernel_size - 1) % 2 == 0
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py:167]:
```suggestion
        assert (kernel_size - 1) % 2 == 0, kernel_size
```

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py [hunk 14]
```diff
@@ -0,0 +1,1886 @@
...
+    """ConvolutionModule.
+
+    Modified from https://github.com/pytorch/audio/blob/main/torchaudio/prototype/models/conv_emformer.py # noqa
+
+    Args:
+      chunk_length (int):
+        Length of each chunk.
+      right_context_length (int):
+        Length of right context.
+      channels (int):
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py:148]: What is the output channel?

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py [hunk 15]
```diff
@@ -0,0 +1,1886 @@
...
+    for b, states in enumerate(state_list[1:], 1):
+        for li, layer in enumerate(states[1]):
+            conv_caches[li].append(layer)
+            if b == batch_size - 1:
+                conv_caches[li] = torch.stack(conv_caches[li], dim=0)
+
+    return [attn_caches, conv_caches]
+
+
+class ConvolutionModule(nn.Module):
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/emformer.py:138]: Are there any tests for classes in this file?

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py (hunk 8 above)
Comment (yaozengwei) [reply egs/librispeech/ASR/conv_emformer_transducer_stateless/test_emformer.py:15]: Checked.

File: egs/librispeech/ASR/conv_emformer_transducer_stateless/stream.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+            )
+
+        self.ground_truth: str = ""
+
+        self.feature: Optional[torch.Tensor] = None
+        # Make sure all feature frames can be used.
+        # Add 2 here since we will drop the first and last after subsampling.
+        self.chunk_length = params.chunk_length
+        self.pad_length = (
+            params.right_context_length + 2 * params.subsampling_factor + 3
```
Comment (pkufool) [inline egs/librispeech/ASR/conv_emformer_transducer_stateless/stream.py:88]: @yaozengwei  I think you should `+7` here, not `+3`,  `self.pad_length` is the frames number before subsampling, we have make sure to keep at least 1 frame after subsampling. Current code would raise errors when running with `max-duration=1`, you can try it.

**Pull request merged**